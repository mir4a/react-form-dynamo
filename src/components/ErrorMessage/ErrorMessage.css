.ErrorMessage {
  position: absolute;
  z-index: 1;
  left: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 10px 20px;
  color: white;
  background-color: var(--error-color);
}

.ErrorMessage-close {
  --btn-size: 20px;
  font-size: 20px;
  font-weight: 700;
  border: none;
  padding: 0;
  line-height: var(--btn-size);
  width: var(--btn-size);
  height: var(--btn-size);
  display: flex;
  align-items: center;
  justify-items: center;
  border-radius: 50%;
  background-color: white;
  color: var(--error-color);
  margin-left: 25px;
}

.ErrorMessage-close:hover {
  cursor: pointer;
  background-color: rgba(255,255,255,0.75);
}

.ErrorMessage-close:focus {
  outline: none;
}

.ErrorMessage-close--text {
  align-self: center;
  flex-grow: 1;
}

.ErrorMessage-animation-enter {
  opacity: 0;
  transform: translateY(-80px);
}

.ErrorMessage-animation-enter-active {
  opacity: 1;
  transform: translateY(0);
  transition: opacity 300ms, transform 300ms;
}

.ErrorMessage-animation-exit {
  opacity: 1;
}

.ErrorMessage-animation-exit-active {
  opacity: 0;
  transform: translateY(-100px);
  transition: opacity 300ms, transform 300ms;
}