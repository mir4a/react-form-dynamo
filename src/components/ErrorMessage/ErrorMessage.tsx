import React, { useState, useEffect, useMemo, useCallback } from 'react';
import { CSSTransition } from 'react-transition-group';
import './ErrorMessage.css';

interface Props {
  message?: string;
  show: boolean;
}

export default React.memo(function ErrorMessage(props: Props) {
  const { message, show } = props;
  const [isShow, setIsShow] = useState(show);
  const [lastMessage, setLastMessage] = useState(message);

  useEffect(() => {
    setIsShow(show);
    if (show) {
      if (message !== lastMessage) {
        setIsShow(true);
      }
      setLastMessage(message);
    }
  }, [message, show, lastMessage]);

  const handleDismiss = useCallback(() => {
    setIsShow(false);
  }, []);

  const renderMessage = useMemo(() => {
    return (
      <>
        <span>{lastMessage}</span>
        <button className="ErrorMessage-close" onClick={handleDismiss}>
          <span className="ErrorMessage-close--text">&times;</span>
        </button>
      </>
    );
  }, [lastMessage, handleDismiss]);

  return (
    <CSSTransition
      in={isShow}
      timeout={200}
      classNames="ErrorMessage-animation"
      unmountOnExit
    >
      <div className="ErrorMessage">{renderMessage}</div>
    </CSSTransition>
  );
});
