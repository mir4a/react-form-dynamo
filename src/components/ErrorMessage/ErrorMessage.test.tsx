import React from 'react';
import renderer from 'react-test-renderer';
import ErrorMessage from './ErrorMessage';

it('renders correctly', () => {
  const tree = renderer
    .create(<ErrorMessage show={true} message="error" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders without message', () => {
  const tree = renderer.create(<ErrorMessage show={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('does not render with show=false', () => {
  const tree = renderer.create(<ErrorMessage show={false} />).toJSON();
  expect(tree).toMatchSnapshot();
});
