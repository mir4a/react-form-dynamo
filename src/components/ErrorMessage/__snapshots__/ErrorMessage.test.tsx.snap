// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render with show=false 1`] = `null`;

exports[`renders correctly 1`] = `
<div
  className="ErrorMessage"
>
  <span>
    error
  </span>
  <button
    className="ErrorMessage-close"
    onClick={[Function]}
  >
    <span
      className="ErrorMessage-close--text"
    >
      ×
    </span>
  </button>
</div>
`;

exports[`renders without message 1`] = `
<div
  className="ErrorMessage"
>
  <span />
  <button
    className="ErrorMessage-close"
    onClick={[Function]}
  >
    <span
      className="ErrorMessage-close--text"
    >
      ×
    </span>
  </button>
</div>
`;
